rules/firestore.json

```diff
@@ -127,60 +126,78 @@ service cloud.firestore {
+            let playerZeroTimeUnchanged = preservesField('remainingMsForZero');
+            let safeTimeIncreasePlayerZero = userIsPlayerOne(part) && safeTimeIncrease(part.playerZero, 'remainingMsForZero', part, update);
+            let playerZeroTimeDecreased = update.remainingMsForZero < part.remainingMsForZero;
+            let playerOneTimeUnchanged = preservesField('remainingMsForOne');
+            let safeTimeIncreasePlayerOne = userIsPlayerZero(part) && safeTimeUpdate(part.playerOne, 'remainingMsForOne', part, update);
+            let playerOneTimeDecreased = update.remainingMsForOne < part.remainingMsForOne;
+            let safeTimeManagement = (playerZeroTimeUnchanged || safeTimeIncreasePlayerZero || playerZeroTimeDecreased) &&
+                (playerOneTimeUnchanged || safeTimeIncreasePlayerOne || playerOneTimeDecreased);
             return isValidPartUpdateDuringGame(part) &&
-                safeTimeUpdatePlayerZero && safeTimeUpdatePlayerOne &&
+                safeTimeManagement &&
                 (preservesField('turn') || safeTurnDecrease(part.turn, update.turn)) &&
                 onlyAcceptsDrawIfProposed(part, update) &&
+//                onlyAcceptsOrRefusesTakeBackIfAsked(part, update) &&
+                onlyResignsInTheirName(part, update) &&
                 safeLastUpdate(part, update);
         }
         function onlyAcceptsDrawIfProposed(part, update) {
             // 6 = draw agreed by zero
             // 7 = draw agreed by one
-            let acceptsDraw = hasField(update, 'result') &&
-                ((userIsPlayerZero(part) && update.result == 6) ||
-                 (userIsPlayerOne(part) && update.result == 7));
+            let updateAcceptsDraw = hasField(update, 'result') &&
+                ((userIsPlayerZero(part) && part.result != 6 && update.result == 6) ||
+                 (userIsPlayerOne(part) && part.result != 7 && update.result == 7));
             let drawWasProposed = hasField(part, 'request') && part.request.code == 'DrawProposed';
-            return acceptsDraw == false ||
-                (acceptsDraw && drawWasProposed);
+            return updateAcceptsDraw == false ||
+                (updateAcceptsDraw && drawWasProposed);
+        }
+//        function onlyAcceptsOrRefusesTakeBackIfAsked(part, update) {
+//            let takeBackWasProposed = hasField(part, 'request') && part.request.code == 'TakeBackAsked';
+//            let updateAnswersToTakeBack = hasField(update, 'request') &&
+//                (part.request.code == 'TakeBackAccepted' || part.request.code == 'TakeBackRefused');
+//            let updateFromDifferentPlayer = hasField(part, 'request') && hasField(update, 'request') &&
+//                part.request.player != update.request.player;
+//            return takeBackWasProposed == false ||
+//                (takeBackWasProposed && updateAnswersToTakeBack && updateFromDifferentPlayer);
+//        }
```

Comment:
why commented ?

Reply:
part.request.player != update.request.player c'est rapide à check, mais, à ton check ces variables là déjà !?

Reply:
c'est quoi part.request.player ?
Sinon c'est commenté car c'est un des check trop couteux laissés sur le côté pour le moment du coup

Reply:
checker que l'auteur de la requête n'est pas l'accepteur de la requête, j'imagine que je demandais ?

Sinon, sans être de tant de l'evil-early-optimisation, faire que dès qu'un truc est faux on retourne, ça diminuerais du calcul, c'est plus un changement de coding pattern qu'un early-optim; non :D ????
Am I the source of all evil by asking that :D ?

Reply:
ah oui ça a du sens. Mais c'est commenté dans tous les cas.

Retourner dès qu'un truc est faux, tu veux dire utiliser `||` et `&&` comme on fait déjà ? `a && b` retourne faux si a est faux sans évaluer b. On stocke des choses temporaires dans des variables mais ça aurait du sens que firestore optimise cela comme il faut (pas évaluer les let non nécessaire). Souviens-toi qu'il y a pas de if dans ce langage aussi, donc je vois pas comment tu peux « retourner plus tôt » autrement :D

Reply:
si tu crois que ce qu'on assigne dans des variables est évalué qu'au return, alors y a rien à faire oui, t'es sûr ou "tu crois qu'si c'est bien fait ils ont optimisé" ?

Reply:
J'ai testé et je confirme !
En gros, ceci ne génère pas d'erreur et retourne bien false:
```
function foo() {
  let x = false;
  let y = resource.bli.bla.i.dont.exist.and.should.throw == true;
  return x && y;
}
```